### DISCUSSION TITLE
How to use different font weights on groupbox widget?

### DISCUSSION THREAD

**Discussion by abrarbinkabir** [qtile/qtile#4478] [Q&A]
I can't use any other font weights in the **groupbox** widget other than the normal one. Is there an option for using other weights or this is the only option available?
`widget.GroupBox(
                font= 'JetBrainsMono Nerd Font Propo',
                fontsize=14,
                padding=8,
                highlight_method = "line",
                highlight_color=theme["lavender"],
                block_highlight_text_color=theme["background"],
                borderwidth = 0,
				active = theme["foreground"],
				inactive = theme["overlay1"],
				disable_drag=True,
),`

**Comment by jrwrigh** [comment]
Short answer: Assuming by "weights" you mean font styles (Bold, Italic, etc.), there is not currently.

Much longer answer:
*This is the result of a deep dive into source code and some light Googling. It may be incorrect, but I'm also partially just documenting this for my own sake*

### What controls the fonts in qtile
Qtile *appears to* use [pango](https://pango.gnome.org/) to handle it's fonts, or at least it's font descriptions. Qtile uses a "custom" interface to the Pango C-library (as [described in code here](https://github.com/qtile/qtile/blob/ad5d0538c51c7ba57ef8520b45b2728fe6caa3a3/libqtile/pangocffi.py#L24-L47)) using the [C Foreign Function Interface (CFFI)](https://cffi.readthedocs.io/en/stable/). That works by writing out definitions of C symbols (functions, enums, etc) of the library that you want to make available in Python (this is done in [`pango_ffi.py`](https://github.com/qtile/qtile/blob/master/libqtile/pango_ffi.py)). Those definitions are then "compiled" at build time.

The font description that you give to `widget.GroupBox` is eventually passed to [`TextLayout`](https://github.com/qtile/qtile/blob/ad5d0538c51c7ba57ef8520b45b2728fe6caa3a3/libqtile/backend/base/drawer.py#L378-L394), which creates a [`FontDescription`](https://docs.gtk.org/Pango/struct.FontDescription.html) object (C struct) which describes the font to be used.

### Why can't we set font style
Pango has a function to set the style of a `FontDescription` called [`pango_font_description_set_style`](https://docs.gtk.org/Pango/method.FontDescription.set_style.html). However, that function is not specified in [`pango_ffi.py`](https://github.com/qtile/qtile/blob/master/libqtile/pango_ffi.py), thus it isn't available as a Python-accessible command.

### How to fix it
First, we'd need to add the [`pango_font_description_set_style`](https://docs.gtk.org/Pango/method.FontDescription.set_style.html) function signature to [`pango_ffi.py`](https://github.com/qtile/qtile/blob/master/libqtile/pango_ffi.py). Then it'd be some simple Python work to add a `font_style` argument to functions and plumb that into `TextLayout`.

**Reply by abrarbinkabir** [reply]
Thanks @jrwrigh for taking the time to answer my questions (also in the mailing list). really appreciate it!
